precompiles/erc20/approve_test.go

```diff
@@ -0,0 +1,562 @@
+	method := s.precompile.Methods[authorization.IncreaseAllowanceMethod]
+	amount := int64(100)
+	increaseAmount := int64(200)
+
+	testcases := []struct {
+		name        string
+		malleate    func() []interface{}
+		postCheck   func()
+		expPass     bool
+		errContains string
+	}{
+		{
+			name:        "fail - empty args",
+			malleate:    func() []interface{} { return nil },
+			errContains: "invalid number of arguments",
+		},
+		{
+			name: "fail - invalid number of arguments",
+			malleate: func() []interface{} {
+				return []interface{}{
+					1, 2, 3,
+				}
+			},
+			errContains: "invalid number of arguments",
+		},
+		{
+			name: "fail - invalid address",
+			malleate: func() []interface{} {
+				return []interface{}{
+					"invalid address", big.NewInt(2),
+				}
+			},
+			errContains: "invalid address",
+		},
+		{
+			name: "fail - invalid amount",
+			malleate: func() []interface{} {
+				return []interface{}{
+					s.keyring.GetAddr(1), "invalid amount",
+				}
+			},
+			errContains: "invalid amount",
+		},
+		{
+			name: "fail - negative amount",
+			malleate: func() []interface{} {
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(-1),
+				}
+			},
+			errContains: "cannot increase allowance with non-positive values",
+		},
+		{
+			name: "pass - increase allowance without existing authorization",
+			malleate: func() []interface{} {
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(increaseAmount),
+				}
+			},
+			expPass: true,
+			postCheck: func() {
+				s.requireSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetAccAddr(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, increaseAmount)),
+					[]string{},
+				)
+			},
+		},
+		{
+			name: "pass - increase allowance with existing authorization",
+			malleate: func() []interface{} {
+				s.setupSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetPrivKey(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, amount)),
+				)
+
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(increaseAmount),
+				}
+			},
+			expPass: true,
+			postCheck: func() {
+				s.requireSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetAccAddr(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, amount+increaseAmount)),
+					[]string{},
+				)
+			},
+		},
+		{
+			name: "pass - increase allowance with existing authorization in different denomination",
+			malleate: func() []interface{} {
+				s.setupSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetPrivKey(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.bondDenom, amount)),
+				)
+
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(increaseAmount),
+				}
+			},
+			expPass: true,
+			postCheck: func() {
+				s.requireSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetAccAddr(0),
+					// NOTE: The approval in the different denomination is overwritten by the
+					// approval for the passed token denomination.
```

Comment:
Do we want this behavior? When "increasing" an approval for a specific denom a prior approval for a different denom is overwritten

Reply:
Ideally not because this creates a weird dynamic where increasing allowance works as an approval as well but I'm not sure if we can work around the way authorizations work here. Maybe we just fail if the `denom` is not the one present on the present authorization?

Reply:
it shouldn't be overwritten but added to the send authorization